src/general/InputEnvironment.cs

```diff
@@ -0,0 +1,40 @@
+using Godot;
+
+/// <summary>
+///     Handles key input for an environment
+/// </summary>
+/// <typeparam name="T">The environment to handle</typeparam>
+public abstract class InputEnvironment<T> : Node
```

Comment:
I don't find the name `Environment`  intuitive in this context. The doc on this class also doesn't help what this is from.
But if I was able to figure out correctly that this is just a little helper meant for all Nodes that accept input, I think this should be named: `NodeInputHelper` and if possible this should actually inherit from `T` instead of directly from Node to make this usable for other Godot Node derived types that want to use this.

Reply:
sadly, you cannot inherit from generic types

Reply:
Then there is not a clean way to make the input thing work for all classes. At that point I think just doing without this intermediate class is the better design. As if InputEnvironment were implemented with composition the main class would anyway need to hookup the focus loss and input events anyway, at that point there might as well be the input group directly in that class.

Reply:
The Class is currently used in the MicrobeEditor and in the MicrobeStage
![image](https://user-images.githubusercontent.com/30809803/96440766-b7587c80-1208-11eb-8261-77532978e408.png)
![image](https://user-images.githubusercontent.com/30809803/96440892-dfe07680-1208-11eb-82d7-509ae391f891.png)
The PlayerMicrobeEditorInput node has PlayerMicrobeEditorInput.cs attached
The PlayerMicrobeInput node has PlayerMicrobeInput.cs attached

Reply:
As I mentioned before, I don't especially like the design where every time you want to accept input you need to add a new Node to it as input. The reason it was added for the microbe input is because it doesn't really fit anywhere else as the player Microbe object gets destroyed and stuff. In the editor having the editor handle the keypresses is pretty nice. Otherwise you'll just need to pipe things through an extra class that makes things less clear in my opinion.

Reply:
you want me to copy all of that input code instead of using some abstractation?

Reply:
It's just 2 methods that need to be duplicated, right? `_Notification` and `_UnhandledInput`? If you want those few lines could be put in a helper class if code duplication need to be reduced? Like how the `TutorialHelper` class is used.

Reply:
This honestly does not feel like a proactive... With this solution, whenever we want to create a new stage we just need to create a new input class, add a node and we can get input via code in a nice way.

Reply:
Input is not limited to stages, with this approach every single place that wants to use the system, needs to add an extra node of indirection, this adds stuff in a bunch of places:
```
src/tutorial/microbe_stage/MicrobeMovement.cs
89:            if (Input.IsActionPressed("g_move_forward"))
94:            if (Input.IsActionPressed("g_move_left"))
99:            if (Input.IsActionPressed("g_move_right"))
104:            if (Input.IsActionPressed("g_move_backwards"))

src/general/PauseMenu.cs
74:        if (@event.IsActionPressed("ui_cancel"))
87:        else if (@event.IsActionPressed("help"))

src/gui_common/TransitionManager.cs
42:        if (@event.IsActionPressed("ui_cancel") && HasQueuedTransitions)

src/gui_common/QuickLoadHandler.cs
16:        if (@event.IsActionPressed("quick_load"))

src/engine/ScreenShotTaker.cs
27:        if (@event.IsActionPressed("screenshot"))

src/engine/FPSCounter.cs
18:        if (@event.IsActionPressed("toggle_FPS"))

src/microbe_stage/MicrobeStage.cs
403:        if (@event.IsActionPressed("g_quick_save"))

src/microbe_stage/MicrobeCamera.cs
109:        if (@event.IsActionPressed("g_zoom_in", true))
114:        if (@event.IsActionPressed("g_zoom_out", true))

```

I want those places to be able to use the new input system *without* adding 8 new Node derived input classes. That's why I really dislike the design where every single place that wants input, would need an intermediate class just to pass along some callbacks to the real code.

Could you weigh in, @Kasterisk ?

Reply:
But... I don't add 8 new Nodes? I just have two nodes: one in the microbeStage and one in the microbeEditor

Reply:
But you are designing a general component to sit between each input group and whatever it is controlling. If this general component is accepted, I think it should be actually general (so no reliance on intermediate Nodes everywhere), otherwise it has no business existing. If it isn't meant to be general then I don't see a reason why InputEnvironment class should exist at all, if it is meant just for the editor input to go through it.

Reply:
I really don't get what intermediate Nodes you are talking about. You add one Node to the Stage and add one class to the node. The class inherits from InputEnviroment.

Reply:
Take the FPSCounter or the ScreenShotTaker for example, for them to use this system they need a new separate Node like ScreenShotTakerInput, that only has the job of calling a method in ScreenShotTaker that input was received. That is completely unnecessary and will clutter the code base with unnecessary files in my opinion. And that's why I don't think this is a good design.

Reply:
Oh, I think I get what you mean... Was a bit stupid there. But still, I think the GUI should notify the Code and the code should not ask the gui for inputs. I mostly come from desktop development, here, this is the correct way to do this, but I think the same patterns apply here.

Reply:
We don't really have a much of a MVC style separation anyway. I don't see the point of for each thing receiving input to add in extra classes. I like that we can have input as close as possible to the actual logic to make it easier to follow.
I've had to explain like 3-4 times on discord how the player is controlled. First explaining the separate input object and then talking about how it goes through some stuff and ends up in Microbe.
I think any input helper code to be added needs to be usable so that for example that FPSCounter class can be cleanly implemented with it, and for that design *requiring* the separate Node derived class, is unnecessary complexity.

Reply:
What about a compromise: for small things like an fpscounter we use the input. methods and for stuff that heavily relies on input we use the InputManager. I know that two different possibilities are confusing, but I think this is the most readable solution

Reply:
I don't like that solution. C#'s language limitations are a bit annoying, but I think in this case they just simply make `InputEnvironment` as designed here unusable. Instead all the stuff in it should be put in a helper class that can be used to get by with very little code duplication for classes that want to use input.

Reply:
I got it! Methods with custom attributes! I think this is a great way to implement this!

Reply:
This attribute will have the input map string and and activation method (press, hold, release) as its ctor

Reply:
Then you can write something like this:
```
[OnInput("e_reset", InputActivationType.Press)]
public void OnReset() {

}
```